openedx/core/djangoapps/call_stack_manager/tests.py

```diff
@@ -0,0 +1,230 @@
+        """ Test for parameterized @donottrack
+        1. Should not log calls of classes specified in the decorator @donotrack
+        """
+        donottrack_child_func()
+        self.assertEqual(ModelMixinCSM, log_capt.call_args[0][1])
+
+    @patch('openedx.core.djangoapps.call_stack_manager.core.log.info')
+    def test_nested_parameterized_donottrack(self, log_capt):
+        """ Tests parameterized nested @donottrack
+        1. should not track call of classes specified in decorated with scope bounded to the respective class
+        """
+        ModelAnotherCSM(id_field=1).save()
+        donottrack_parent_func()
+        self.assertEqual(ModelAnotherCSM, log_capt.call_args_list[0][0][1])
+        self.assertEqual(ModelMixinCSM, log_capt.call_args_list[1][0][1])
+
+    @patch('openedx.core.djangoapps.call_stack_manager.core.log.info')
+    def test_nested_parameterized_donottrack_after(self, log_capt):
+        """ Tests parameterized nested @donottrack
+        1. should not track call of classes specified in decorated with scope bounded to the respective function
+        """
+        donottrack_child_func()
+        # class with CallStackManager as Manager
+        ModelAnotherCSM(id_field=1).save()
+        # test is this- that this should get called.
+        ModelAnotherCSM.objects.filter(id_field=1)
+        self.assertEqual(ModelMixinCSM, log_capt.call_args_list[0][0][1])
+        self.assertEqual(ModelAnotherCSM, log_capt.call_args_list[1][0][1])
+
+    @patch('openedx.core.djangoapps.call_stack_manager.core.log.info')
+    def test_donottrack_called_in_func(self, log_capt):
+        """ test for function which calls decorated function.
+        """
+        ModelAnotherCSM(id_field=1).save()
+        ModelMixinCSM(id_field=1).save()
+        track_it()
+        self.assertEqual(ModelMixinCSM, log_capt.call_args_list[0][0][1])
+        self.assertEqual(ModelAnotherCSM, log_capt.call_args_list[1][0][1])
+        self.assertEqual(ModelMixinCSM, log_capt.call_args_list[2][0][1])
+        self.assertEqual(ModelAnotherCSM, log_capt.call_args_list[3][0][1])
+
+    @patch('openedx.core.djangoapps.call_stack_manager.core.log.info')
+    def test_donottrack_child_too(self, log_capt):
+        """
+        1. subclass should not be tracked when superclass is called in a @donottrack decorated function
+        """
+        ModelWithCSM(id_field=1).save()
+        ModelWithCSMChild(id_field=1, id1_field=1).save()
+        abstract_do_not_track()
+        self.assertEqual(len(log_capt.call_args_list), 0)
+
+    @patch('openedx.core.djangoapps.call_stack_manager.core.log.info')
+    def test_duplication(self, log_capt):
+        """ Test for duplication of call stacks
+        1. no duplication of call stacks
+        """
+        for dummy in range(1, 5):
+            ModelMixinCSM(id_field=1).save()
+        self.assertEqual(len(log_capt.call_args_list), 1)
+
+
+@donottrack(ModelWithCSMChild)
```

Comment:
Nit: I prefer the actual tests to be at the bottom of the file, with the supporting code above.